### PR TITLE
Improve test_namedtensor.py with named tensor equality check

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24087 Name inference for softmax, log_softmax and Dimname overloads.
* #24028 Fix out= function semantics for named tensors.
* #23973 Update tensor.view_names / tensor.names_ API
* #23962 Rename set_names -> view_names, set_names_ -> names_
* #23804 Implement tensor.align_to(names), torch.align_tensors(t1, t2, ...)
* #23808 Named inference for contiguous(), bernoulli variants, and dropout.
* #23746 Implement name inference rule for empty_like, clone
* #23743 Add `names` argument to ones, rand, randn, zeros, full; fix empty
* **#23801 Improve test_namedtensor.py with named tensor equality check**
* #23624 Add name propagation for at::alias, add tensor.set_names

Test Plan
- Code reading. assertTensorDataAndNamesEqual isn't used in this commit
but it'll be used in future commits.
- [namedtensor ci]

gh-metadata: pytorch pytorch 23801 gh/zou3519/90/head

Differential Revision: [D16667816](https://our.internmc.facebook.com/intern/diff/D16667816)

### CODE REVIEW

File: test/test_namedtensor.py [hunk 1]
```diff
@@ -26,6 +26,14 @@ class TestNamedTensor(TestCase):
     def test_trivial(self):
         pass
 
+    # TODO(rzou): Some form of this check should be added to self.assertEqual.
+    # Right now I don't know what it should look like.
+    def assertTensorDataAndNamesEqual(self, x, y):
+        self.assertEqual(x.names, y.names)
+        unnamed_x = x.set_names(None)
+        unnamed_y = y.set_names(None)
+        self.assertEqual(unnamed_x, unnamed_y)
```
Comment (nairbv) [inline test/test_namedtensor.py]: if `assertEqual` doesn't check names, why do we need to make these unnamed?
if `assertEqual` does compare names, why do we need this new method?

Should this check for named tensors being enabled in the config before calling set_names?

I think I didn't realize `set_names` was returning a new tensor instead of mutating the passed tensor, which was why I was suggesting `names_`. Still not sure what we should be calling it though, if it's not a mutator then `set` doesn't seem right either.

Comment (zou3519) [reply test/test_namedtensor.py]: > if assertEqual doesn't check names, why do we need to make these unnamed?
if assertEqual does compare names, why do we need this new method?

`assertEqual` doesn't support named tensors because it calls some operations that don't have named inference rules implemented. In the long run it will support named tensors and it should probably check names.

The reason why I am unsure is because it does some strange behavior in other cases. One example is `assertEqual(torch.ones(3, dtype=torch.float), torch.ones(3, dtype=torch.int))` is True.

> Should this check for named tensors being enabled in the config before calling set_names?

I don't think so. This check is exclusive to TestNamedTensor, which can't be run unless named tensors are enable.

> I think I didn't realize set_names was returning a new tensor instead of mutating the passed tensor, which was why I was suggesting names_. Still not sure what we should be calling it though, if it's not a mutator then set doesn't seem right either.

Yeah, this question has been on my mind for a while. `names_` is a good name for the mutator. However, we need a name for the out-of-place variant.

**Review by nairbv — APPROVED** [review]
